FileSystemException: File name too long

When running a sufficiently nested workflow, I get the error `File name too long`, because WDL is generating a temporary file named `temp-s3-6826696772619511254cromwell-execution_best_practise_3bf9c436-31f7-4d86-ba87-cf54248e05bc_call-ConvertPairedFastQsToUnmappedBamWf_unmap.ConvertPairedFastQsToUnmappedBamWf_e10704e8-7ff4-4e4c-89ad-48270734ba25_call-CreateFoFN_write_lines_238fd975f9e3d166dbec07b20ad88c51.tmp`, which is greater than 256 characters and thus too large for any Linux system.

I'm running Cromwell 36 on with the AWS Batch backend. The traceback is:

```
2018-10-19 01:29:11,310 cromwell-system-akka.dispatchers.backend-dispatcher-8681 ERROR - Failed command instantiation
java.lang.Exception: Failed command instantiation
        at cromwell.backend.standard.StandardAsyncExecutionActor.instantiatedCommand(StandardAsyncExecutionActor.scala:565)
        at cromwell.backend.standard.StandardAsyncExecutionActor.instantiatedCommand$(StandardAsyncExecutionActor.scala:500)
        at cromwell.backend.impl.aws.AwsBatchAsyncBackendJobExecutionActor.instantiatedCommand$lzycompute(AwsBatchAsyncBackendJobExecutionActor.scala:74)
        at cromwell.backend.impl.aws.AwsBatchAsyncBackendJobExecutionActor.instantiatedCommand(AwsBatchAsyncBackendJobExecutionActor.scala:74)  
        at cromwell.backend.standard.StandardAsyncExecutionActor.commandScriptContents(StandardAsyncExecutionActor.scala:313)
        at cromwell.backend.standard.StandardAsyncExecutionActor.commandScriptContents$(StandardAsyncExecutionActor.scala:312)
        at cromwell.backend.impl.aws.AwsBatchAsyncBackendJobExecutionActor.commandScriptContents(AwsBatchAsyncBackendJobExecutionActor.scala:74)
        at cromwell.backend.impl.aws.AwsBatchAsyncBackendJobExecutionActor.batchJob$lzycompute(AwsBatchAsyncBackendJobExecutionActor.scala:132) 
        at cromwell.backend.impl.aws.AwsBatchAsyncBackendJobExecutionActor.batchJob(AwsBatchAsyncBackendJobExecutionActor.scala:131)
        at cromwell.backend.impl.aws.AwsBatchAsyncBackendJobExecutionActor.executeAsync(AwsBatchAsyncBackendJobExecutionActor.scala:342)
        at cromwell.backend.standard.StandardAsyncExecutionActor.executeOrRecover(StandardAsyncExecutionActor.scala:943)
        at cromwell.backend.standard.StandardAsyncExecutionActor.executeOrRecover$(StandardAsyncExecutionActor.scala:935)
        at cromwell.backend.impl.aws.AwsBatchAsyncBackendJobExecutionActor.executeOrRecover(AwsBatchAsyncBackendJobExecutionActor.scala:74)
        at cromwell.backend.async.AsyncBackendJobExecutionActor.$anonfun$robustExecuteOrRecover$1(AsyncBackendJobExecutionActor.scala:65)
        at cromwell.core.retry.Retry$.withRetry(Retry.scala:38)
        at cromwell.backend.async.AsyncBackendJobExecutionActor.withRetry(AsyncBackendJobExecutionActor.scala:61)
        at cromwell.backend.async.AsyncBackendJobExecutionActor.cromwell$backend$async$AsyncBackendJobExecutionActor$$robustExecuteOrRecover(AsyncBackendJobExecutionActor.scala:65)
        at cromwell.backend.async.AsyncBackendJobExecutionActor$$anonfun$receive$1.applyOrElse(AsyncBackendJobExecutionActor.scala:88)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:172)
        at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:172)
        at akka.actor.Actor.aroundReceive(Actor.scala:517)
        at akka.actor.Actor.aroundReceive$(Actor.scala:515)
        at cromwell.backend.impl.aws.AwsBatchAsyncBackendJobExecutionActor.aroundReceive(AwsBatchAsyncBackendJobExecutionActor.scala:74)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)
        at akka.actor.ActorCell.invoke(ActorCell.scala:557)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
        at akka.dispatch.Mailbox.run(Mailbox.scala:225)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
        at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: common.exception.AggregatedMessageException: Error(s):
Could not evaluate expression: write_lines(array_of_files): [Attempted 1 time(s)] - FileSystemException: /tmp/temp-s3-6826696772619511254cromwell-execution_best_practise_3bf9c436-31f7-4d86-ba87-cf54248e05
bc_call-ConvertPairedFastQsToUnmappedBamWf_unmap.ConvertPairedFastQsToUnmappedBamWf_e10704e8-7ff4-4e4c-89ad-48270734ba25_call-CreateFoFN_write_lines_238fd975f9e3d166dbec07b20ad88c51.tmp: File name too lon
g
        at common.validation.Validation$ValidationTry$.toTry$extension1(Validation.scala:68)
        at common.validation.Validation$ValidationTry$.toTry$extension0(Validation.scala:64)
        at cromwell.backend.standard.StandardAsyncExecutionActor.instantiatedCommand(StandardAsyncExecutionActor.scala:563)
        ... 31 common frames omitted
```

It seems that each "segment" of this filename, e.g. `bc_call-ConvertPairedFastQsToUnmappedBamWf_unmap.ConvertPairedFastQsToUnmappedBamWf_e10704e8-7ff4-4e4c-89ad-48270734ba25`, should probably be in a different folder, not just as a different segment of the same filename.

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3Channel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3Channel.java
@@ -6,6 +6,6 @@ import java.nio.file.Path;
 
 public interface S3Channel {
     default Path createTempFile(S3Path path) throws IOException {
-        return Files.createTempFile("", "");
+        return Files.createTempFile(path.getFileName().toString(), "");
     }
 }

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3Channel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3Channel.java
@@ -5,6 +5,12 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 
 public interface S3Channel {
+    /**
+     * Create a temporary place for the S3 file on the local filesystem, with the s3 filename mixed in
+     * e.g. "/tmp/script-tempSuffix"
+     * @param path S3 path to create a file for
+     * @return Path on the local filesystem to the temporary file that was created
+     */
     default Path createTempFile(S3Path path) throws IOException {
         return Files.createTempFile(path.getFileName().toString(), "");
     }

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3Channel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3Channel.java
@@ -1,0 +1,11 @@
+package org.lerch.s3fs;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public interface S3Channel {
+    default Path createTempFile(S3Path path) throws IOException {
+        return Files.createTempFile("", "");
+    }
+}

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3FileChannel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3FileChannel.java
@@ -23,7 +23,7 @@ import software.amazon.awssdk.services.s3.model.S3Object;
 
 import static java.lang.String.format;
 
-public class S3FileChannel extends FileChannel {
+public class S3FileChannel extends FileChannel implements S3Channel {
 
     private S3Path path;
     private Set<? extends OpenOption> options;
@@ -42,7 +42,7 @@ public class S3FileChannel extends FileChannel {
                 !this.options.contains(StandardOpenOption.CREATE))
             throw new NoSuchFileException(format("target not exists: %s", path));
 
-        tempFile = Files.createTempFile("", "");
+        tempFile = createTempFile(path);
         boolean removeTempFile = true;
         try {
             if (exists) {

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3FileChannel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3FileChannel.java
@@ -42,7 +42,7 @@ public class S3FileChannel extends FileChannel {
                 !this.options.contains(StandardOpenOption.CREATE))
             throw new NoSuchFileException(format("target not exists: %s", path));
 
-        tempFile = Files.createTempFile("temp-s3-", key.replaceAll("/", "_"));
+        tempFile = Files.createTempFile("", "");
         boolean removeTempFile = true;
         try {
             if (exists) {

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
@@ -2,7 +2,10 @@ package org.lerch.s3fs;
 
 import static java.lang.String.format;
 
-import java.io.*;
+import java.io.ByteArrayInputStream;
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
 import java.nio.ByteBuffer;
 import java.nio.channels.SeekableByteChannel;
 import java.nio.file.*;

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
@@ -23,7 +23,7 @@ import software.amazon.awssdk.services.s3.model.GetObjectRequest;
 import software.amazon.awssdk.services.s3.model.PutObjectRequest;
 import software.amazon.awssdk.services.s3.model.S3Object;
 
-public class S3SeekableByteChannel implements SeekableByteChannel {
+public class S3SeekableByteChannel implements SeekableByteChannel, S3Channel {
 
     private S3Path path;
     private Set<? extends OpenOption> options;
@@ -49,7 +49,7 @@ public class S3SeekableByteChannel implements SeekableByteChannel {
                 !this.options.contains(StandardOpenOption.CREATE))
             throw new NoSuchFileException(format("target not exists: %s", path));
 
-        tempFile = Files.createTempFile("", "");
+        tempFile = createTempFile(path);
         boolean removeTempFile = true;
         try {
             if (exists) {

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
@@ -49,10 +49,7 @@ public class S3SeekableByteChannel implements SeekableByteChannel {
                 !this.options.contains(StandardOpenOption.CREATE))
             throw new NoSuchFileException(format("target not exists: %s", path));
 
-        Path tempdir = Files.createTempDirectory(null);
-        Path tempSubdir = tempdir.resolve(path.getParent());
-        Files.createDirectories(tempSubdir);
-        tempFile = Files.createTempFile(tempSubdir, "temp-s3-" + path.getFileName().toString(), null);
+        tempFile = Files.createTempFile("", "");
         boolean removeTempFile = true;
         try {
             if (exists) {

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
@@ -2,10 +2,7 @@ package org.lerch.s3fs;
 
 import static java.lang.String.format;
 
-import java.io.ByteArrayInputStream;
-import java.io.BufferedInputStream;
-import java.io.IOException;
-import java.io.InputStream;
+import java.io.*;
 import java.nio.ByteBuffer;
 import java.nio.channels.SeekableByteChannel;
 import java.nio.file.*;

--- a/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
+++ b/filesystems/s3/src/main/java/org/lerch/s3fs/S3SeekableByteChannel.java
@@ -49,7 +49,10 @@ public class S3SeekableByteChannel implements SeekableByteChannel {
                 !this.options.contains(StandardOpenOption.CREATE))
             throw new NoSuchFileException(format("target not exists: %s", path));
 
-        tempFile = Files.createTempFile("temp-s3-", key.replaceAll("/", "_"));
+        Path tempdir = Files.createTempDirectory(null);
+        Path tempSubdir = tempdir.resolve(path.getParent());
+        Files.createDirectories(tempSubdir);
+        tempFile = Files.createTempFile(tempSubdir, "temp-s3-" + path.getFileName().toString(), null);
         boolean removeTempFile = true;
         try {
             if (exists) {

--- a/src/ci/bin/testCentaurAws.sh
+++ b/src/ci/bin/testCentaurAws.sh
@@ -48,7 +48,6 @@ cromwell::build::run_centaur \
     -e draft3_optional_input_from_scatter \
     -e iwdr_input_string \
     -e globbingindex \
-    -e file_name_too_long \
     -e cwl_cache_between_workflows \
     -e abort.scheduled_abort \
     -e cwl_cache_within_workflow \

